File: server/modules/copilot/bankSnapshot.test.ts
import assert from 'node:assert/strict';
import { describe, it } from 'node:test';
import type { NessieSnapshot } from '../nessie/types.ts';
import type { WorkspaceState } from '../../store/userStore.ts';
import { bankSnapshotKey, toBankSnapshot } from './bankSnapshot.ts';

function workspace(balancesApplied: boolean): WorkspaceState {
  return {
    mode: 'sandbox',
    businessName: 'Juniper Kitchen',
    nessieCustomerId: 'cust-1',
    accountIds: ['acc-op', 'acc-card'],
    merchants: { m1: { name: 'Sysco', category: 'Food & Beverage' } },
    balancesApplied,
    provisionedAt: '2026-09-12T00:00:00.000Z',
  };
}

function snapshot(): NessieSnapshot {
  return {
    customer: { _id: 'cust-1', first_name: 'Jordan', last_name: 'Rivera', address: { street_number: '1', street_name: 'Main', city: 'Houston', state: 'TX', zip: '77001' } },
    accounts: [
      { _id: 'acc-op', type: 'Checking', nickname: 'Operating', rewards: 0, balance: 1000, customer_id: 'cust-1' },
      { _id: 'acc-card', type: 'Credit Card', nickname: 'Card', rewards: 0, balance: 300, customer_id: 'cust-1' },
    ],
    deposits: [
      { _id: 'dep-1', type: 'deposit', transaction_date: '2026-09-10', status: 'completed', payee_id: 'acc-op', medium: 'balance', amount: 500, description: 'Client payment · Acme' },
      { _id: 'dep-2', type: 'deposit', transaction_date: '2026-09-30', status: 'pending', payee_id: 'acc-op', medium: 'balance', amount: 4000, description: 'Invoice 1041 · Client A · net 30' },
    ],
    withdrawals: [{ _id: 'wd-1', type: 'withdrawal', transaction_date: '2026-09-08', status: 'completed', payer_id: 'acc-op', medium: 'balance', amount: 200, description: 'Payroll · Gusto' }],
    purchases: [{ _id: 'pur-1', type: 'merchant', merchant_id: 'm1', payer_id: 'acc-card', purchase_date: '2026-09-09', amount: 50, status: 'completed', medium: 'balance', description: 'Weekly produce' }],
    bills: [{ _id: 'bill-1', status: 'recurring', payee: 'Harbor Property Group', nickname: 'Office rent', payment_date: '2026-08-01', recurring_date: 1, upcoming_payment_date: '2026-10-01', payment_amount: 2200, account_id: 'acc-op' }],
    transfers: [],
    merchants: { m1: { name: 'Sysco', category: 'Food & Beverage' } },
  };
}

describe('toBankSnapshot', () => {
  it('derives deposit-account balances from settled history when the bank did not apply it, and leaves cards as owed', () => {
    const push = toBankSnapshot(workspace(false), snapshot());
    assert.equal(push.customerId, 'cust-1');
    assert.deepEqual(push.accounts.map((a) => [a.id, a.balance]), [['acc-op', 1300], ['acc-card', 300]]);
  });

  it('keeps the bank balances when the bank applied the history itself', () => {
    const push = toBankSnapshot(workspace(true), snapshot());
    assert.deepEqual(push.accounts.map((a) => a.balance), [1000, 300]);
  });

  it('carries merchant names, calendar dates and the bill schedule the ledger normalises', () => {
    const push = toBankSnapshot(workspace(true), snapshot());
    assert.deepEqual(push.purchases[0], { id: 'pur-1', accountId: 'acc-card', date: '2026-09-09', status: 'completed', amount: 50, description: 'Weekly produce', merchantName: 'Sysco', merchantCategory: 'Food & Beverage' });
    assert.equal(push.deposits[1]?.status, 'pending');
    assert.equal(push.withdrawals[0]?.description, 'Payroll · Gusto');
    assert.deepEqual(push.bills[0], { id: 'bill-1', accountId: 'acc-op', payee: 'Harbor Property Group', nickname: 'Office rent', paymentDate: '2026-10-01', status: 'recurring', amount: 2200, recurring: true });
  });

  it('fingerprints the snapshot so an unchanged one is not pushed twice', () => {
    const a = bankSnapshotKey(toBankSnapshot(workspace(true), snapshot()));
    const b = bankSnapshotKey(toBankSnapshot(workspace(true), snapshot()));
    assert.equal(a, b);
    const changed = snapshot();
    changed.deposits[0]!.amount = 501;
    assert.notEqual(bankSnapshotKey(toBankSnapshot(workspace(true), changed)), a);
  });
});

File: services/ledger-service/src/test/java/com/cashflowcopilot/bank/BankSnapshotTest.java
package com.cashflowcopilot.bank;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;

/**
 * The BFF pushes the workspace's bank snapshot; the ledger becomes the same business the Keel
 * dashboard shows, and available cash comes from the ingested balances rather than a live call.
 */
@SpringBootTest(properties = {"app.demo-mode=true", "app.database-url=", "app.internal-service-token=bank-test"})
@AutoConfigureMockMvc
class BankSnapshotTest {

    @Autowired MockMvc mvc;
    @Autowired ObjectMapper mapper;
    @Autowired BankAccountRepository bankAccounts;

    MockHttpServletRequestBuilder as(String subject, MockHttpServletRequestBuilder request) {
        return request.header("X-Internal-Service-Token", "bank-test").header("X-Auth-Subject", subject);
    }

    JsonNode ok(String subject, MockHttpServletRequestBuilder request) throws Exception {
        return mapper.readTree(mvc.perform(as(subject, request))
                .andExpect(status().isOk()).andReturn().getResponse().getContentAsString());
    }

    void verify(String subject) throws Exception {
        ok(subject, post("/v1/persona/status").contentType("application/json").content("{\"status\":\"approved\"}"));
    }

    static String day(int offset) {
        return LocalDate.now(ZoneOffset.UTC).plusDays(offset).toString();
    }

    /** Two accounts, five money movements, an internal transfer and two bills: the Keel workspace in miniature. */
    static String workspaceSnapshot(String customerId) {
        return """
                {
                  "customerId": "%s",
                  "accounts": [
                    {"id": "acc-op", "type": "Checking", "nickname": "Operating", "balance": 9000.00},
                    {"id": "acc-card", "type": "Credit Card", "nickname": "Business card", "balance": 1500.00}
                  ],
                  "deposits": [
                    {"id": "dep-1", "accountId": "acc-op", "date": "%s", "status": "completed", "amount": 6000.00, "description": "Client payment · Acme"},
                    {"id": "dep-2", "accountId": "acc-op", "date": "%s", "status": "pending", "amount": 4000.00, "description": "Invoice 1041 · Client A · net 30"},
                    {"id": "dep-3", "accountId": "acc-res", "date": "%s", "status": "completed", "amount": 500.00, "description": "Transfer · Reserve sweep"}
                  ],
                  "withdrawals": [
                    {"id": "wd-1", "accountId": "acc-op", "date": "%s", "status": "completed", "amount": 5400.00, "description": "Payroll · Gusto"}
                  ],
                  "purchases": [
                    {"id": "pur-1", "accountId": "acc-card", "date": "%s", "status": "completed", "amount": 320.00, "description": "Weekly produce", "merchantName": "Sysco", "merchantCategory": "Food & Beverage"}
                  ],
                  "bills": [
                    {"id": "bill-rent", "accountId": "acc-op", "payee": "Harbor Property Group", "nickname": "Office rent", "paymentDate": "%s", "status": "recurring", "amount": 2200.00, "recurring": true},
                    {"id": "bill-ins", "accountId": "acc-op", "payee": "Northline Insurance", "nickname": "Liability insurance", "paymentDate": "%s", "status": "pending", "amount": 1000.00, "recurring": false}
                  ]
                }
                """.formatted(customerId, day(-3), day(-12), day(-8), day(-5), day(-2), day(-20), day(5));
    }

    @Test
    void aPushedSnapshotBecomesTheUsersCashPositionAndLedger() throws Exception {
        String subject = "google:bank-push-user";
        verify(subject);

        JsonNode result = ok(subject, post("/v1/bank/snapshot")
                .contentType("application/json").content(workspaceSnapshot("cust-a")));
        // The internal transfer is not cash flow; everything else is one event each.
        assertThat(result.get("insertedEvents").asInt()).isEqualTo(6);
        assertThat(result.get("updatedEvents").asInt()).isZero();

        JsonNode summary = ok(subject, get("/v1/accounts/summary"));
        assertThat(summary.get("totalBalance").decimalValue()).isEqualByComparingTo("7500.00");
        assertThat(summary.get("accounts")).hasSize(2);
        assertThat(summary.get("lastSyncedAt").isNull()).isFalse();

        JsonNode forecast = ok(subject, get("/v1/forecast"));
        assertThat(forecast.get("currentCash").decimalValue()).isEqualByComparingTo("7500.00");

        JsonNode dashboard = ok(subject, get("/v1/dashboard"));
        assertThat(dashboard.get("totals").get("availableCash").decimalValue()).isEqualByComparingTo("7500.00");
        List<String> overdue = new ArrayList<>();
        dashboard.get("overdueReceivables").forEach(r -> overdue.add(r.get("counterpartyLabel").asText()));
        assertThat(overdue).containsExactly("Client A");
        List<String> obligations = new ArrayList<>();
        dashboard.get("upcomingObligations").forEach(o -> obligations.add(o.get("label").asText()));
        assertThat(obligations).contains("Harbor Property Group", "Northline Insurance");

        List<String> categories = new ArrayList<>();
        ok(subject, get("/v1/cash-events")).forEach(e -> categories.add(e.get("category").asText()));
        assertThat(categories).contains("payroll", "food_beverage", "client_payment", "client_invoice");
    }

    @Test
    void aSecondPushIsIdempotentAndAChangedCustomerReplacesTheOldFeed() throws Exception {
        String subject = "google:bank-switch-user";
        verify(subject);
        ok(subject, post("/v1/bank/snapshot").contentType("application/json").content(workspaceSnapshot("cust-a")));

        JsonNode again = ok(subject, post("/v1/bank/snapshot")
                .contentType("application/json").content(workspaceSnapshot("cust-a")));
        assertThat(again.get("insertedEvents").asInt()).isZero();
        assertThat(again.get("updatedEvents").asInt()).isEqualTo(6);

        String other = """
                {"customerId": "cust-b",
                 "accounts": [{"id": "acc-x", "type": "Checking", "nickname": "Main", "balance": 1234.00}],
                 "deposits": [{"id": "dep-x", "accountId": "acc-x", "date": "%s", "status": "completed", "amount": 10.00, "description": "Sales"}],
                 "withdrawals": [], "purchases": [], "bills": []}
                """.formatted(day(-1));
        ok(subject, post("/v1/bank/snapshot").contentType("application/json").content(other));

        assertThat(ok(subject, get("/v1/accounts/summary")).get("totalBalance").decimalValue()).isEqualByComparingTo("1234.00");
        assertThat(ok(subject, get("/v1/cash-events"))).hasSize(1);
    }

    @Test
    void requiresAVerifiedIdentityAndAWellFormedSnapshot() throws Exception {
        String subject = "google:bank-unverified-user";
        mvc.perform(as(subject, post("/v1/bank/snapshot")
                        .contentType("application/json").content(workspaceSnapshot("cust-a"))))
                .andExpect(status().isForbidden());

        verify(subject);
        mvc.perform(as(subject, post("/v1/bank/snapshot")
                        .contentType("application/json").content("{\"customerId\":\"c\",\"accounts\":[{\"id\":\"a\",\"type\":\"Checking\",\"balance\":\"lots\"}]}")))
                .andExpect(status().isBadRequest());
    }

    @Test
    void seedingWithoutBankDataAddsOnlyTheVendorHistory() throws Exception {
        String subject = "google:bank-history-only-user";
        verify(subject);

        JsonNode seed = ok(subject, post("/v1/demo/seed")
                .contentType("application/json").content("{\"includeBankData\":false}"));
        assertThat(seed.get("seeded").asBoolean()).isTrue();
        assertThat(seed.get("invoices").asInt()).isEqualTo(12);
        assertThat(seed.get("bankEvents").asInt()).isZero();
        assertThat(seed.get("ledgerEvents").asInt()).isZero();
        assertThat(seed.get("todos").asInt()).isZero();
        assertThat(ok(subject, get("/v1/cash-events"))).isEmpty();
        assertThat(ok(subject, get("/v1/forecast")).get("currentCash").decimalValue()).isZero();
    }

    @Test
    void aUserSeededBeforeBalancesWereStoredGetsThemBackfilled() throws Exception {
        String subject = "google:bank-backfill-user";
        verify(subject);
        ok(subject, post("/v1/demo/seed"));
        assertThat(ok(subject, get("/v1/accounts/summary")).get("totalBalance").decimalValue())
                .isEqualByComparingTo("8000.00");

        // Simulate the pre-upgrade state: events present, balances table empty.
        bankAccounts.replaceAll(userId(subject), List.of(), java.time.Instant.now());
        assertThat(ok(subject, get("/v1/forecast")).get("currentCash").decimalValue()).isZero();

        JsonNode again = ok(subject, post("/v1/demo/seed"));
        assertThat(again.get("seeded").asBoolean()).isFalse();
        assertThat(ok(subject, get("/v1/forecast")).get("currentCash").decimalValue())
                .isEqualByComparingTo("8000.00");
    }

    java.util.UUID userId(String subject) throws Exception {
        return java.util.UUID.fromString(ok(subject, get("/v1/me")).get("userId").asText());
    }

    @Test
    void theDemoCustomerIsAlwaysServedFromTheFixture() throws Exception {
        String subject = "google:bank-demo-fixture-user";
        verify(subject);
        JsonNode sync = ok(subject, post("/v1/nessie/sync"));
        assertThat(sync.get("insertedEvents").asInt()).isGreaterThan(0);
        assertThat(ok(subject, get("/v1/accounts/summary")).get("totalBalance").decimalValue())
                .isEqualByComparingTo("8000.00");
    }
}
